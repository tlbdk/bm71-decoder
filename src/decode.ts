// Links
// https://ww1.microchip.com/downloads/en/DeviceDoc/BM70-71-Bluetooth-Low-Energy-Module-Users-Guide-DS50002542.pdf

// Request format
// Header(Start(0xAA), Length(0x0001-0xFFFF)), Opcode(0x00-0xFF), Data(0-n bytes), Checksum(0x00)

// Response format
// Header(Start(0xAA), Length(0x0001-0xFFFF)), Data(0-n bytes), Checksum(0x00)

export enum CommandOpcode {
  READ_LOCAL_INFORMATION = 0x01,
  RESET = 0x02,
  STATUS_READ = 0x03,
  ADC_READ = 0x04,
  SHUTDOWN = 0x05,
  DEBUG = 0x06,
  NAME_READ = 0x07,
  NAME_WRITE = 0x08,
  PDL_ERASE = 0x09,
  PAIR_MODE_READ = 0x0a,
  PAIR_MODE_WRITE = 0x0b,
  PDL_READ = 0x0c,
  DEVICE_ERASE = 0x0d,
  DIO_CONTROL = 0x0e,
  PWM_CONTROL = 0x0f,
  RSSI_READ = 0x10,
  ADV_DATA_WRITE = 0x11,
  SCAN_DATA_WRITE = 0x12,
  ADV_PARAM_SET = 0x13,
  SCAN_PARAM_SET = 0x15,
  SCAN_ENABLE_SET = 0x16,
  CONNECT = 0x17,
  CONNECT_CANCEL = 0x18,
  CONNECTION_PARAM_UPDATE = 0x19,
  DISCONNECT = 0x1b,
  SET_ADV_ENABLE = 0x1c,
  REMOTE_NAME_READ = 0x1f,
  CLIENT_WRITE_REQUEST_PREPARE = 0x20,
  CLIENT_WRITE_REQUEST_EXECUTE = 0x21,
  CLIENT_BLOB_REQUEST_READ = 0x22,
  SERVER_WRITE_RESPONSE_PREPARE = 0x28,
  SERVER_WRITE_RESPONSE_EXECUTE = 0x29,
  SERVER_BLOB_RESPONSE_READ = 0x2a,
  CLIENT_HANDLE_VALUE_CONFIRM = 0x2d,
  CLIENT_DISCOVER_ALL_SERVICES = 0x30,
  CLIENT_DISCOVER_CHARACTERISTICS = 0x31,
  CLIENT_CHARACTERISTIC_READ = 0x32,
  CLIENT_CHARACTERISTIC_UUID_READ = 0x33,
  CLIENT_CHARACTERISTIC_WRITE = 0x34,
  TRANSPARENT_ENABLE = 0x35,
  SERVER_ERROR_RESPONSE = 0x37,
  SERVER_CHARACTERISTIC_SEND = 0x38,
  SERVER_CHARACTERISTIC_UPDATE = 0x39,
  SERVER_CHARACTERISTIC_READ = 0x3a,
  SERVER_ALL_SERVICES_READ = 0x3b,
  SERVER_SERVICE_READ = 0x3c,
  SERVER_WRITE_RESPONSE_SEND = 0x3d,
  SERVER_READ_RESPONSE_SEND = 0x3e,
  TRANSPARENT_DATA_SEND = 0x3f,
  PASSKEY_ENTRY = 0x40,
  PASSKEY_YES_NO_CONFIRM = 0x41,
  PAIRING_REQUEST = 0x42,
  CONFIG_MODE_CLOSE = 0x52,
  RECEIVER_TEST = 0x53,
  TRANSMITTER_TEST = 0x54,
  END_TEST = 0x55
}

export enum CommandStatusOpcode {
  COMMAND_SUCCEEDED = 0x00,
  UNKNOWN_COMMAND = 0x01,
  UNKNOWN_CONNECTION_IDENTIFIER = 0x02,
  HARDWARE_FAILURE = 0x03,
  AUTHENTICATION_FAILURE = 0x05,
  PIN_OR_KEY_MISSING = 0x06,
  MEMORY_CAPACITY_EXCEEDED = 0x07,
  CONNECTION_TIMEOUT = 0x08,
  CONNECTION_LIMIT_EXCEEDED = 0x09,
  ACL_CONNECTION_ALREADY_EXISTS = 0x0b,
  COMMAND_DISALLOWED = 0x0c,
  CONNECTION_REJECTED_DUE_TO_LIMITED_RESOURCES = 0x0d,
  CONNECTION_REJECTED_DUE_TO_SECURITY_REASONS = 0x0e,
  CONNECTION_REJECTED_DUE_TO_UNACCEPTABLE_BD_ADDR = 0x0f,
  CONNECTION_ACCEPT_TIMEOUT_EXCEEDED = 0x10,
  UNSUPPORTED_FEATURE_OR_PARAMETER_VALUE = 0x11,
  INVALID_COMMAND_PARAMETERS = 0x12,
  REMOTE_USER_TERMINATED_CONNECTION = 0x13,
  REMOTE_DEVICE_TERMINATED_CONNECTION_DUE_TO_LOW_RESOURCES = 0x14,
  REMOTE_DEVICE_TERMINATED_CONNECTION_DUE_TO_POWER_OFF = 0x15,
  CONNECTION_TERMINATED_BY_LOCAL_HOST = 0x16,
  PAIRING_NOT_ALLOWED = 0x18,
  UNSPECIFIED_ERROR = 0x1f,
  INSTANT_PASSED = 0x28,
  PAIRING_WITH_UNIT_KEY_NOT_SUPPORTED = 0x29,
  INSUFFICIENT_SECURITY = 0x2f,
  CONNECTION_REJECTED_DUE_TO_NO_SUITABLE_CHANNEL_FOUND = 0x39,
  CONTROLLER_BUSY = 0x3a,
  UNACCEPTABLE_CONNECTION_INTERVAL = 0x3b,
  DIRECTED_ADVERTISING_TIMEOUT = 0x3c,
  CONNECTION_TERMINATED_DUE_TO_MIC_FAILURE = 0x3d,
  CONNECTION_FAILED_TO_BE_ESTABLISHED = 0x3e,
  INVALID_OFFSET = 0x77,
  INVALID_HANDLE = 0x81,
  READ_NOT_PERMITTED = 0x82,
  WRITE_NOT_PERMITTED = 0x83,
  INVALID_PDU = 0x84,
  INSUFFICIENT_AUTHENTICATION = 0x85,
  REQUEST_NOT_SUPPORTED = 0x86,
  INSUFFICIENT_AUTHORIZATION = 0x88,
  PREPARE_QUEUE_FULL = 0x89,
  ATTRIBUTE_NOT_FOUND = 0x8a,
  ATTRIBUTE_NOT_LONG = 0x8b,
  INSUFFICIENT_ENCRYPTION_KEY_SIZE = 0x8c,
  INVALID_ATTRIBUTE_VALUE_LENGTH = 0x8d,
  UNLIKELY_ERROR = 0x8e,
  INSUFFICIENT_ENCRYPTION = 0x8f,
  UNSUPPORTED_GROUP_TYPE = 0x90,
  INSUFFICIENT_RESOURCES = 0x91,
  APPLICATION_DEFINED_ERROR = 0xf0,
  UART_CHECKSUM_ERROR = 0xff
}

export interface ReadLocalInformationCommand {
  opcode: 0x01
}

export interface UnknownCommand {
  data: Buffer
}

export type Command = ReadLocalInformationCommand | UnknownCommand

export enum EventOpcode {
  EVENT_NONE = 0x00,
  PASSKEY_REQUEST = 0x60,
  PAIR_COMPLETE = 0x61,
  PASSKEY_YES_NO_REQUEST = 0x62,
  ADVERTISING_REPORT = 0x70,
  LE_CONNECT_COMPLETE = 0x71,
  DISCONNECT_COMPLETE = 0x72,
  CONNECTION_PARAMTER_UPDATE = 0x73,
  SPP_CONNECT_COMPLETE = 0x74,
  COMMAND_COMPLETE = 0x80,
  STATUS_REPORT = 0x81,
  LE_END_TEST_RESULT = 0x82,
  CONFIGURE_MODE_STATUS = 0x8f,
  CLIENT_DISCOVER_ALL_SERVICES_RESULT = 0x90,
  CLIENT_DISCOVER_CHARACTERISTICS_RESULT = 0x91,
  CLIENT_DISCOVER_CHARACTERISTICS_DESCRIPTORS_RESULT = 0x92,
  CLIENT_CHARACTERISTIC_VALUE_RECEIVED = 0x93,
  SERVER_CHARACTERISTIC_VALUE_WRITE = 0x98,
  SERVER_CHARACTERISTIC_VALUE_READ = 0x99,
  TRANSPARENT_DATA_RECEIVED = 0x9a,
  ERROR = 0xff
}

export interface CommandCompleteEvent {
  opcode: EventOpcode.COMMAND_COMPLETE
  commandOpcode: CommandOpcode
  completionStatus: CommandStatusOpcode
}

export interface ReadLocalInformationResponse extends CommandCompleteEvent {
  type: 'read_local_information'
  commandOpcode: CommandOpcode.READ_LOCAL_INFORMATION
  firmwareVersion: number
  macAddress: string
}

export interface UnknownCommandCompleteEvent extends CommandCompleteEvent {
  type: 'unknown_command_complete'
  data: Buffer
}

export enum PacketType {
  CONNECTABLE_UNDIRECTED_ADVERTISING = 0x00,
  CONNECTABLE_DIRECTED_ADVERTISING = 0x01,
  SCANNABLE_UNDIRECTED_ADVERTISING = 0x02,
  NON_CONNECTABLE_UNDIRECTED_ADVERTISING = 0x03,
  SCAN_RESPONSE = 0x04
}

export enum AddressType {
  PUBLIC = 0x00,
  RANDOM = 0x01
}

export interface AdvertisingReportEvent {
  type: 'advertising_report'
  opcode: EventOpcode.ADVERTISING_REPORT
  packetType: string
  addressType: string
  macAddress: string
  payload: Buffer
  rssi: number
}

export interface ShortEvent {
  type: 'short'
  data: Buffer
}

interface UnknownEvent {
  type: 'unknown'
  data: Buffer
}

type Event =
  | ReadLocalInformationResponse
  | UnknownCommandCompleteEvent
  | AdvertisingReportEvent
  | ShortEvent
  | UnknownEvent

function sliceToNextStart(data: Buffer, offset = 0): Buffer {
  const nextAA = data.indexOf(0xaa, offset)
  if (nextAA > 0) {
    return data.slice(nextAA)
  } else {
    return Buffer.alloc(0)
  }
}

export function decodeEvents(data: Buffer): Event[] {
  const results: Event[] = []
  while (data.length > 5) {
    // Start byte is always 0xAA
    const startByte = data.readUInt8(0)
    if (startByte !== 0xaa) {
      // Skip to next 0xAA if start byte is wrong
      data = sliceToNextStart(data)
      continue
    }

    const dataLength = data.readUInt16BE(1)
    if (dataLength === 0 || dataLength + 3 > data.length) {
      // Length should always be 1 or larger as opcode is also counted
      data = sliceToNextStart(data, 1)
      continue
    }

    if (dataLength + 4 < data.length && data.readUInt8(dataLength + 4) !== 0xaa) {
      // The next message does not start with 0xAA so this message is probably short
      results.push({
        type: 'short',
        data: data.slice(0, dataLength)
      })
      data = sliceToNextStart(data, 1)
      continue
    }

    // TODO Validate checksum
    const checksum = data.readUInt8(dataLength + 3)

    // Event opcode is the first param
    const opcode = data.readUInt8(3)
    switch (opcode) {
      case EventOpcode.COMMAND_COMPLETE: {
        const commandOpcode = data.readUInt8(4)
        switch (commandOpcode) {
          case CommandOpcode.READ_LOCAL_INFORMATION: {
            results.push({
              type: 'read_local_information',
              opcode: EventOpcode.COMMAND_COMPLETE,
              commandOpcode: CommandOpcode.READ_LOCAL_INFORMATION,
              completionStatus: data.readUInt8(5),
              firmwareVersion: data.readUInt32BE(6),
              macAddress: data.slice(10, 16).toString('hex')
            })
            break
          }
          default: {
            results.push({
              type: 'unknown_command_complete',
              opcode: EventOpcode.COMMAND_COMPLETE,
              commandOpcode: CommandOpcode.READ_LOCAL_INFORMATION,
              completionStatus: data.readUInt8(5),
              data: data.slice(0, dataLength + 3)
            })
          }
        }
        break
      }
      case EventOpcode.ADVERTISING_REPORT: {
        const payloadLength = data.readUInt8(12)
        if (13 + payloadLength + 1 > data.length) {
          // Payload length should be shorter than data else skip to next record
          data = sliceToNextStart(data, 1)
          continue
        }

        results.push({
          type: 'advertising_report',
          opcode: EventOpcode.ADVERTISING_REPORT,
          packetType: PacketType[data.readUInt8(4)],
          addressType: AddressType[data.readUInt8(5)],
          macAddress: data.slice(6, 12).toString('hex'),
          payload: data.slice(13, 13 + payloadLength),
          rssi: data.readInt8(13 + payloadLength + 1)
        })
        break
      }
      default: {
        data = sliceToNextStart(data, 1)
        continue
      }
    }

    const newData = data.slice(dataLength + 4)
    data = newData
  }

  return results
}

export function decodeCommands(data: Buffer): Command[] {
  const results: Command[] = []
  while (data.length > 5) {
    // Start byte is always 0xAA
    const startByte = data.readUInt8(0)
    if (startByte !== 0xaa) {
      // Skip to next 0xAA if start byte is wrong
      data = sliceToNextStart(data)
      continue
    }

    const dataLength = data.readUInt16BE(1)
    if (dataLength === 0) {
      // Length should always be 1 or larger as we need at least one param
      data = sliceToNextStart(data, 1)
      continue
    }

    // TODO: Validate if dataLength is too large
    // TODO Validate checksum
    const checksum = data.readUInt8(dataLength + 3)

    const opcode = data.readUInt8(3)

    switch (opcode) {
      case CommandOpcode.READ_LOCAL_INFORMATION: {
        results.push({
          opcode: 0x01
        })
        break
      }
      default: {
        results.push({
          data: data.slice(0, dataLength + 3)
        })
      }
    }
  }
  return results
}

export function encode(opcode: CommandOpcode): Buffer {
  return Buffer.alloc(0)
}

export function checksum(data: Buffer): Buffer {
  return Buffer.alloc(0)
}
