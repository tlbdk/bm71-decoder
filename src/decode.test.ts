import { AdvertisingReportEvent, decodeEvents } from './decode'

describe('decode', () => {
  describe('decodeEvent', () => {
    it('should decode response for READ_LOCAL_INFORMATION', () => {
      const events = decodeEvents(Buffer.from('AA000E800100AAAAAAAAFFFFFFFFFFFF01FF', 'hex'))
      expect(events).toMatchSnapshot()
    })

    it('should decode ADVERTISING_REPORT event', () => {
      const events = decodeEvents(
        Buffer.from('aa001e7000012730bb4ce9731302011a020a080cff4c0010073f1f7d9e650118c648', 'hex')
      )
      expect(events).toMatchSnapshot()
    })

    it('should decode many ADVERTISING_REPORT event', () => {
      const events = decodeEvents(
        Buffer.from(
          [
            'aa001e7000012730bb4ce9731302011a020a080cff4c0010073f1f7d9e650118c648',
            'aa000b7004012730bb4ce97300c600',
            'aa002a7003018bb2d6e54a3c1f1eff060001092002c370885c7b74bfeb1523b70c1cdd8f6c0fc1f1d40464fbae31',
            'aa0029700001789c482c62611e0201061aff4c000c0e00844861db7a0681c1c8527e938610054a1cfade6dad8d',
            'aa00197000019f66173bf6740e0201060aff4c001005411c38492ba388',
            'aa0025700000c2b94a29e8181a02010611061745561073ed09a6b34e7ce552',
            'aa001d700400c2b94a29e8181209162a2518e8294ab9c107161820c0a8fe32ba6d',
            'aa002a700301b18b5a325b061f02011a03036ffd17166ffd037e630c6dc980d348892ff5a591d2c91ba49c91bc0b',
            'aa00117000011bfb7848e20806c29584481877bc40',
            'aa000b700401b0a729b9cb5c00a977',
            'aa000b700401c2958448187700a925',
            'aa00117000013858eb15735806789c482c6261b022',
            'aa0028700000bf8a7992bf701d02011a07030f180a18fffe11094a616272612045766f6c766520373565ab34',
            'aa0011700001e1b16801114b068b47b9c9c77eb2d6',
            'aa001c700001321b54f0d7441102011a020a0c0aff4c0010050318bcefc9a2e6',
            'aa00277002014a4e52d342601c03039ffe17169ffe02517a663344714c63775573000001797f7452e2a62e',
            'aa00157004014a4e52d342600a09ffe00006eaca7d3104a514',
            'aa0011700000a23595d0b69c06020102020a02b91f',
            'aa0018700400a23595d0b69c0d0c09736b6e2d6c6170746f7000b902',
            'aa001c700001bb511c0424611102011a020a0c0aff4c0010051b1c81ee0fa5b8',
            'aa0027700300b2f560a468381c1bff750042040180603868a460f5b23a68a460f5b101000000000000a30e',
            'aa002a7003010a3b6e7d97791f1eff060001092006d25714f5a92cbdf7a4f83366af',
            'aa002a700301c3cef39ce04a1f1eff060001092002b9e376356cb0e07a89b57fdc49b0ddbe226d1ff2c5e52fa1a6',
            'aa001c70000122b76024935b1102011a020a0c0aff4c0010051e18ed1b80a218'
          ].join(''),
          'hex'
        )
      )
      const json = JSON.stringify(
        events
          .filter((e): e is AdvertisingReportEvent => e.type === 'advertising_report')
          .map(e => {
            return {
              ...e,
              payload: e.payload.toString('hex')
            }
          }),
        null,
        2
      )
      expect(events).toMatchSnapshot()
    })
  })
})
