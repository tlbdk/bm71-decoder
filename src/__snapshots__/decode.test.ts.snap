// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`decode decodeEvent should decode ADVERTISING_REPORT event 1`] = `
Array [
  Object {
    "addressType": "RANDOM",
    "macAddress": "2730bb4ce973",
    "opcode": 112,
    "packetType": "CONNECTABLE_UNDIRECTED_ADVERTISING",
    "payload": Object {
      "data": Array [
        2,
        1,
        26,
        2,
        10,
        8,
        12,
        255,
        76,
        0,
        16,
        7,
        63,
        31,
        125,
        158,
        101,
        1,
        24,
      ],
      "type": "Buffer",
    },
    "rssi": 72,
    "type": "advertising_report",
  },
]
`;

exports[`decode decodeEvent should decode many ADVERTISING_REPORT event 1`] = `
Array [
  Object {
    "addressType": "RANDOM",
    "macAddress": "2730bb4ce973",
    "opcode": 112,
    "packetType": "CONNECTABLE_UNDIRECTED_ADVERTISING",
    "payload": Object {
      "data": Array [
        2,
        1,
        26,
        2,
        10,
        8,
        12,
        255,
        76,
        0,
        16,
        7,
        63,
        31,
        125,
        158,
        101,
        1,
        24,
      ],
      "type": "Buffer",
    },
    "rssi": 72,
    "type": "advertising_report",
  },
  Object {
    "addressType": "RANDOM",
    "macAddress": "2730bb4ce973",
    "opcode": 112,
    "packetType": "SCAN_RESPONSE",
    "payload": Object {
      "data": Array [],
      "type": "Buffer",
    },
    "rssi": 0,
    "type": "advertising_report",
  },
  Object {
    "addressType": "RANDOM",
    "macAddress": "8bb2d6e54a3c",
    "opcode": 112,
    "packetType": "NON_CONNECTABLE_UNDIRECTED_ADVERTISING",
    "payload": Object {
      "data": Array [
        30,
        255,
        6,
        0,
        1,
        9,
        32,
        2,
        195,
        112,
        136,
        92,
        123,
        116,
        191,
        235,
        21,
        35,
        183,
        12,
        28,
        221,
        143,
        108,
        15,
        193,
        241,
        212,
        4,
        100,
        251,
      ],
      "type": "Buffer",
    },
    "rssi": 49,
    "type": "advertising_report",
  },
  Object {
    "addressType": "RANDOM",
    "macAddress": "789c482c6261",
    "opcode": 112,
    "packetType": "CONNECTABLE_UNDIRECTED_ADVERTISING",
    "payload": Object {
      "data": Array [
        2,
        1,
        6,
        26,
        255,
        76,
        0,
        12,
        14,
        0,
        132,
        72,
        97,
        219,
        122,
        6,
        129,
        193,
        200,
        82,
        126,
        147,
        134,
        16,
        5,
        74,
        28,
        250,
        222,
        109,
      ],
      "type": "Buffer",
    },
    "rssi": -115,
    "type": "advertising_report",
  },
  Object {
    "addressType": "RANDOM",
    "macAddress": "9f66173bf674",
    "opcode": 112,
    "packetType": "CONNECTABLE_UNDIRECTED_ADVERTISING",
    "payload": Object {
      "data": Array [
        2,
        1,
        6,
        10,
        255,
        76,
        0,
        16,
        5,
        65,
        28,
        56,
        73,
        43,
      ],
      "type": "Buffer",
    },
    "rssi": -120,
    "type": "advertising_report",
  },
  Object {
    "data": Object {
      "data": Array [
        170,
        0,
        37,
        112,
        0,
        0,
        194,
        185,
        74,
        41,
        232,
        24,
        26,
        2,
        1,
        6,
        17,
        6,
        23,
        69,
        86,
        16,
        115,
        237,
        9,
        166,
        179,
        78,
        124,
        229,
        82,
        170,
        0,
        29,
        112,
        4,
        0,
      ],
      "type": "Buffer",
    },
    "type": "short",
  },
  Object {
    "addressType": "PUBLIC",
    "macAddress": "c2b94a29e818",
    "opcode": 112,
    "packetType": "SCAN_RESPONSE",
    "payload": Object {
      "data": Array [
        9,
        22,
        42,
        37,
        24,
        232,
        41,
        74,
        185,
        193,
        7,
        22,
        24,
        32,
        192,
        168,
        254,
        50,
      ],
      "type": "Buffer",
    },
    "rssi": 109,
    "type": "advertising_report",
  },
  Object {
    "addressType": "RANDOM",
    "macAddress": "b18b5a325b06",
    "opcode": 112,
    "packetType": "NON_CONNECTABLE_UNDIRECTED_ADVERTISING",
    "payload": Object {
      "data": Array [
        2,
        1,
        26,
        3,
        3,
        111,
        253,
        23,
        22,
        111,
        253,
        3,
        126,
        99,
        12,
        109,
        201,
        128,
        211,
        72,
        137,
        47,
        245,
        165,
        145,
        210,
        201,
        27,
        164,
        156,
        145,
      ],
      "type": "Buffer",
    },
    "rssi": 11,
    "type": "advertising_report",
  },
  Object {
    "addressType": "RANDOM",
    "macAddress": "1bfb7848e208",
    "opcode": 112,
    "packetType": "CONNECTABLE_UNDIRECTED_ADVERTISING",
    "payload": Object {
      "data": Array [
        194,
        149,
        132,
        72,
        24,
        119,
      ],
      "type": "Buffer",
    },
    "rssi": 64,
    "type": "advertising_report",
  },
  Object {
    "addressType": "RANDOM",
    "macAddress": "b0a729b9cb5c",
    "opcode": 112,
    "packetType": "SCAN_RESPONSE",
    "payload": Object {
      "data": Array [],
      "type": "Buffer",
    },
    "rssi": 119,
    "type": "advertising_report",
  },
  Object {
    "addressType": "RANDOM",
    "macAddress": "c29584481877",
    "opcode": 112,
    "packetType": "SCAN_RESPONSE",
    "payload": Object {
      "data": Array [],
      "type": "Buffer",
    },
    "rssi": 37,
    "type": "advertising_report",
  },
  Object {
    "addressType": "RANDOM",
    "macAddress": "3858eb157358",
    "opcode": 112,
    "packetType": "CONNECTABLE_UNDIRECTED_ADVERTISING",
    "payload": Object {
      "data": Array [
        120,
        156,
        72,
        44,
        98,
        97,
      ],
      "type": "Buffer",
    },
    "rssi": 34,
    "type": "advertising_report",
  },
  Object {
    "addressType": "PUBLIC",
    "macAddress": "bf8a7992bf70",
    "opcode": 112,
    "packetType": "CONNECTABLE_UNDIRECTED_ADVERTISING",
    "payload": Object {
      "data": Array [
        2,
        1,
        26,
        7,
        3,
        15,
        24,
        10,
        24,
        255,
        254,
        17,
        9,
        74,
        97,
        98,
        114,
        97,
        32,
        69,
        118,
        111,
        108,
        118,
        101,
        32,
        55,
        53,
        101,
      ],
      "type": "Buffer",
    },
    "rssi": 52,
    "type": "advertising_report",
  },
  Object {
    "addressType": "RANDOM",
    "macAddress": "e1b16801114b",
    "opcode": 112,
    "packetType": "CONNECTABLE_UNDIRECTED_ADVERTISING",
    "payload": Object {
      "data": Array [
        139,
        71,
        185,
        201,
        199,
        126,
      ],
      "type": "Buffer",
    },
    "rssi": -42,
    "type": "advertising_report",
  },
  Object {
    "addressType": "RANDOM",
    "macAddress": "321b54f0d744",
    "opcode": 112,
    "packetType": "CONNECTABLE_UNDIRECTED_ADVERTISING",
    "payload": Object {
      "data": Array [
        2,
        1,
        26,
        2,
        10,
        12,
        10,
        255,
        76,
        0,
        16,
        5,
        3,
        24,
        188,
        239,
        201,
      ],
      "type": "Buffer",
    },
    "rssi": -26,
    "type": "advertising_report",
  },
  Object {
    "addressType": "RANDOM",
    "macAddress": "4a4e52d34260",
    "opcode": 112,
    "packetType": "SCANNABLE_UNDIRECTED_ADVERTISING",
    "payload": Object {
      "data": Array [
        3,
        3,
        159,
        254,
        23,
        22,
        159,
        254,
        2,
        81,
        122,
        102,
        51,
        68,
        113,
        76,
        99,
        119,
        85,
        115,
        0,
        0,
        1,
        121,
        127,
        116,
        82,
        226,
      ],
      "type": "Buffer",
    },
    "rssi": 46,
    "type": "advertising_report",
  },
  Object {
    "addressType": "RANDOM",
    "macAddress": "4a4e52d34260",
    "opcode": 112,
    "packetType": "SCAN_RESPONSE",
    "payload": Object {
      "data": Array [
        9,
        255,
        224,
        0,
        6,
        234,
        202,
        125,
        49,
        4,
      ],
      "type": "Buffer",
    },
    "rssi": 20,
    "type": "advertising_report",
  },
  Object {
    "addressType": "PUBLIC",
    "macAddress": "a23595d0b69c",
    "opcode": 112,
    "packetType": "CONNECTABLE_UNDIRECTED_ADVERTISING",
    "payload": Object {
      "data": Array [
        2,
        1,
        2,
        2,
        10,
        2,
      ],
      "type": "Buffer",
    },
    "rssi": 31,
    "type": "advertising_report",
  },
  Object {
    "addressType": "PUBLIC",
    "macAddress": "a23595d0b69c",
    "opcode": 112,
    "packetType": "SCAN_RESPONSE",
    "payload": Object {
      "data": Array [
        12,
        9,
        115,
        107,
        110,
        45,
        108,
        97,
        112,
        116,
        111,
        112,
        0,
      ],
      "type": "Buffer",
    },
    "rssi": 2,
    "type": "advertising_report",
  },
  Object {
    "addressType": "RANDOM",
    "macAddress": "bb511c042461",
    "opcode": 112,
    "packetType": "CONNECTABLE_UNDIRECTED_ADVERTISING",
    "payload": Object {
      "data": Array [
        2,
        1,
        26,
        2,
        10,
        12,
        10,
        255,
        76,
        0,
        16,
        5,
        27,
        28,
        129,
        238,
        15,
      ],
      "type": "Buffer",
    },
    "rssi": -72,
    "type": "advertising_report",
  },
  Object {
    "addressType": "PUBLIC",
    "macAddress": "b2f560a46838",
    "opcode": 112,
    "packetType": "NON_CONNECTABLE_UNDIRECTED_ADVERTISING",
    "payload": Object {
      "data": Array [
        27,
        255,
        117,
        0,
        66,
        4,
        1,
        128,
        96,
        56,
        104,
        164,
        96,
        245,
        178,
        58,
        104,
        164,
        96,
        245,
        177,
        1,
        0,
        0,
        0,
        0,
        0,
        0,
      ],
      "type": "Buffer",
    },
    "rssi": 14,
    "type": "advertising_report",
  },
  Object {
    "data": Object {
      "data": Array [
        170,
        0,
        42,
        112,
        3,
        1,
        10,
        59,
        110,
        125,
        151,
        121,
        31,
        30,
        255,
        6,
        0,
        1,
        9,
        32,
        6,
        210,
        87,
        20,
        245,
        169,
        44,
        189,
        247,
        164,
        248,
        51,
        102,
        175,
        170,
        0,
        42,
        112,
        3,
        1,
        195,
        206,
      ],
      "type": "Buffer",
    },
    "type": "short",
  },
  Object {
    "addressType": "RANDOM",
    "macAddress": "c3cef39ce04a",
    "opcode": 112,
    "packetType": "NON_CONNECTABLE_UNDIRECTED_ADVERTISING",
    "payload": Object {
      "data": Array [
        30,
        255,
        6,
        0,
        1,
        9,
        32,
        2,
        185,
        227,
        118,
        53,
        108,
        176,
        224,
        122,
        137,
        181,
        127,
        220,
        73,
        176,
        221,
        190,
        34,
        109,
        31,
        242,
        197,
        229,
        47,
      ],
      "type": "Buffer",
    },
    "rssi": -90,
    "type": "advertising_report",
  },
  Object {
    "addressType": "RANDOM",
    "macAddress": "22b76024935b",
    "opcode": 112,
    "packetType": "CONNECTABLE_UNDIRECTED_ADVERTISING",
    "payload": Object {
      "data": Array [
        2,
        1,
        26,
        2,
        10,
        12,
        10,
        255,
        76,
        0,
        16,
        5,
        30,
        24,
        237,
        27,
        128,
      ],
      "type": "Buffer",
    },
    "rssi": 24,
    "type": "advertising_report",
  },
]
`;

exports[`decode decodeEvent should decode response for READ_LOCAL_INFORMATION 1`] = `
Array [
  Object {
    "commandOpcode": 1,
    "completionStatus": 0,
    "firmwareVersion": 2863311530,
    "macAddress": "ffffffffffff",
    "opcode": 128,
    "type": "read_local_information",
  },
]
`;
